cpp/src/bitmask/null_mask.cu

```diff
@@ -330,6 +396,26 @@ rmm::device_buffer bitmask_and(bitmask_type const *mask1, size_type begin_bit1,
   return dest_mask;
 }
 
+// convert [first_bit_index,last_bit_index) to
+// [first_word_index,last_word_index)
+struct to_word_index : public thrust::unary_function<size_type, size_type> {
+  bool _first = false;
+  size_type const* _d_bit_indices = nullptr;
+
+  __host__ to_word_index(bool first, size_type const* d_bit_indices) : _first(first), _d_bit_indices(d_bit_indices) {}
+
+  __device__ size_type operator()(const size_type& i) const {
+    auto bit_index = _d_bit_indices[i];
+    if (_first) {  // inclusive
+      return word_index(bit_index);
+    }
+    else {  // exclusive
```

Comment:
So this step looks to be a result of the fact that in `indices`, the beginning index is inclusive while the end bit index is exclusive. Did you consider doing a pre-processing step of `indices` to convert the end indices to be inclusive? 

Then you wouldn't need a runtime check inside the functor, and a lambda could be used for the offsets iterator instead of this functor.

Reply:
> Did you consider doing a pre-processing step of `indices` to convert the end indices to be inclusive?

Yeah, I was wondering why the following code uses the inclusive index for the last bit (against the convention in libcudf elsewhere).
 
https://github.com/rapidsai/cudf/blob/4a18029c698b66384441ecc4805aa68078e18390/cpp/src/bitmask/null_mask.cu#L123

This can be confusing in one regard, but may be OK as this is low level implementation detail and is not part of the public API. I will fix this.

Reply:
I spent a long time going back and forth on that kernel on whether or not the `last_bit_index` should be inclusive or exclusive. The logic ended up being a lot cleaner and easier making it inclusive. Pretty much the exact same reason why I suggested making it inclusive here :)

Reply:
I tried to implement this,

but I am using cub::DeviceSegmentedReduce::Sum() to compute segmented null counts as suggested in https://github.com/rapidsai/cudf/issues/3600, and cub takes **exclusive** _OffsetIteratorT d_end_offsets_.

So, this requires converting to inclusive and converting back to exclusive again, and this doesn't look pretty to me, and what do you thing about this?